apache/flink
# [FLINK-19712][Coordination] Do not restart CREATED executions in RestartPipelinedRegionFailoverStrategy

## What is the purpose of the change

When a task fails and it is `RestartPipelinedRegionFailoverStrategy`, all tasks in the region of the failed task and in the downstream regions will be canceled for later re-scheduling. However, these tasks can be still in `CREATED` state so that there is no need to cancel these tasks.

The PR skips canceling these tasks which can speed up the failover and reduce a lot of unnecessary CANCELING logs.ELING logs.

## Brief change log

  - refactor builder for `TestingSchedulingExecutionVertex`
  - add state to executions in `RestartPipelinedRegionFailoverStrategyTest`
  - refactor/deduplicate verification logic in `RestartPipelinedRegionFailoverStrategyTest`
  - add test that executions in `CREATED` state do not get restarted
  - fix expected restarts in `BatchFineGrainedRecoveryITCase` because subsequent mappers do not get restarted anymore if their parent mapper fails

## Verifying this change

unit tests
